Content.Shared/Nutrition/EntitySystems/ThirstSystem.cs

```diff
@@ -154,22 +155,25 @@ private void UpdateEffects(EntityUid uid, ThirstComponent component)
 
     public override void Update(float frameTime)
     {
-        _accumulatedFrameTime += frameTime;
+        base.Update(frameTime);
 
-        if (_accumulatedFrameTime > 1)
+        var query = EntityQueryEnumerator<ThirstComponent>();
+        while (query.MoveNext(out var uid, out var thirst))
         {
-            var query = EntityManager.EntityQueryEnumerator<ThirstComponent>();
-            while (query.MoveNext(out var uid, out var comp))
-            {
-                UpdateThirst(comp, - comp.ActualDecayRate);
-                var calculatedThirstThreshold = GetThirstThreshold(comp, comp.CurrentThirst);
-                if (calculatedThirstThreshold != comp.CurrentThirstThreshold)
-                {
-                    comp.CurrentThirstThreshold = calculatedThirstThreshold;
-                    UpdateEffects(uid, comp);
-                }
-            }
-            _accumulatedFrameTime -= 1;
+            if (_timing.CurTime < thirst.NextUpdateTime)
+                continue;
+
+            thirst.NextUpdateTime += thirst.UpdateRate;
```

Comment:
also this should support pause/unpause (just search for PausedEvent or like that). Right now if entity is paused and then unpaused it will make all the updates missed, making entity dead pretty much.